GPU/Common/ReplacedTexture.cpp

```diff
@@ -390,21 +512,28 @@ bool ReplacedTexture::CopyLevelTo(int level, void *out, int rowPitch) {
 		return false;
 	}
 
-	if (rowPitch < info.w * 4) {
-		ERROR_LOG(G3D, "Replacement rowPitch=%d, but w=%d (level=%d)", rowPitch, info.w * 4, level);
-		return false;
-	}
-	_assert_msg_(data.size() == info.w * info.h * 4, "Data has wrong size");
+	if (fmt == Draw::DataFormat::R8G8B8A8_UNORM) {
+		if (rowPitch < info.w * 4) {
+			ERROR_LOG(G3D, "Replacement rowPitch=%d, but w=%d (level=%d)", rowPitch, info.w * 4, level);
+			return false;
+		}
 
-	if (rowPitch == info.w * 4) {
-		ParallelMemcpy(&g_threadManager, out, &data[0], info.w * 4 * info.h);
+		_assert_msg_(data.size() == info.w * info.h * 4, "Data has wrong size");
+
+		if (rowPitch == info.w * 4) {
+			ParallelMemcpy(&g_threadManager, out, &data[0], info.w * 4 * info.h);
+		} else {
+			const int MIN_LINES_PER_THREAD = 4;
+			ParallelRangeLoop(&g_threadManager, [&](int l, int h) {
+				for (int y = l; y < h; ++y) {
+					memcpy((uint8_t *)out + rowPitch * y, &data[0] + info.w * 4 * y, info.w * 4);
+				}
+				}, 0, info.h, MIN_LINES_PER_THREAD);
+		}
 	} else {
-		const int MIN_LINES_PER_THREAD = 4;
-		ParallelRangeLoop(&g_threadManager, [&](int l, int h) {
-			for (int y = l; y < h; ++y) {
-				memcpy((uint8_t *)out + rowPitch * y, &data[0] + info.w * 4 * y, info.w * 4);
-			}
-			}, 0, info.h, MIN_LINES_PER_THREAD);
+		// TODO: Add sanity checks here for other formats?
+		// Just gonna do a memcpy, slightly scared of the parallel ones.
```

Comment:
It may cause some contention but it should be fine for large amounts of data.  If the game is showing one large 512x512 texture as a background, and it's upscaled to 2048x2048, the parallel is likely to be worth it.  At most, could adjust the heuristics on when.

It would mostly not be worth it if we had a ton of textures to load and it starved them.

-[Unknown]

Reply:
Fair enough, but with the compressed texture formats, the data size is drastically smaller, 1 byte per pixel instead of four. But yeah, I'll have another look at the heuristics and probably activate here too.

Reply:
I think it's still very likely worth it for copying 2MB (0.5 of 2048x2048.)  Ideal might be to only use a few threads for it, but especially on a device with 12+ threads, it's practically wasting them not to...

-[Unknown]